File: 05_binary_search/01_bin_search.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        // Binary Search
        int i = -1;
        int start = 0, end = nums.size() - 1;

        while (start <= end)
        {
            int mid = start + (end - start) / 2; // for preventing overflow
            // mid = (start+end)/2
            if (nums[mid] == target)
            {
                i = mid;
                break;
            }
            else if (nums[mid] < target)
            {
                start = mid + 1;
            }
            else
            {
                end = mid - 1;
            }
        }
        return i;
    }
};

File: 05_binary_search/01_bin_search.py
from typing import List
#https://docs.python.org/3.6/library/typing.html
#refer to link for typing

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        #list is sorted
        low = 0
        high = len(nums)-1
        ans = -1

        while(low <= high):
            mid = int((low+high)/2)
            if(nums[mid] == target):
                ans = mid
                break
            elif (nums[mid] < target):
                low = mid+1
            else:
                high = mid-1
        
        return ans
